### PR TITLE
Policy document tests

### PR DESCRIPTION
**Pull request by Briansbum**
As I mentioned in my [issue](https://github.com/gruntwork-io/terratest/issues/120) on the subject I find myself doing the simple task of verifying that policy documents exist.

With a view to getting going with tests on IAM/S3 permissions these functions will let you check for the existence of policy documents and get their contents.

### CODE REVIEW

File: modules/aws/iam.go [hunk 1]
```diff
@@ -133,6 +133,50 @@ func EnableMfaDeviceE(t *testing.T, iamClient *iam.IAM, mfaDevice *iam.VirtualMF
...
+func AssertIamPolicyExistsE(t *testing.T, region string, policyARN string) error {
+	iamClient, err := NewIamClientE(t, region)
+	if err != nil {
+		return err
+	}
+
+	_, err = iamClient.GetPolicyVersion(&iam.GetPolicyVersionInput{
+		PolicyArn: &policyARN,
+	})
+	return err
```
Comment (brikis98) [inline modules/aws/iam.go]: Instead of duplicating this, couldn't this method call `GetIamPolicyDocumentE` and discard the contents? Same for the S3 methods.

**Comment by Briansbum** [comment]
I've added a test for each of IAM and S3 which makes use of the main functions I've made.

Also in there is a new bit of functionality I realised was going to be useful when I was writing the tests so you can feed a policy document, from terraform output or whatever, and a policy ARN/Bucket Name into the `AssertXPolicyIsCorrect` to test that AWS hasn't fiddled with the document at all which has happened to me enough that I want to test it.

[Also here are the tests passing 👍](https://gist.github.com/Briansbum/f51d0d8cc807b2988acbe3c5a2f3edad)

File: modules/aws/iam.go [hunk 2]
```diff
@@ -133,6 +138,97 @@ func EnableMfaDeviceE(t *testing.T, iamClient *iam.IAM, mfaDevice *iam.VirtualMF
...
+	dst := new(bytes.Buffer)
+
+	src := []byte(policyDocument)
+
+	err := json.Compact(dst, src)
+	if err != nil {
+		return err
+	}
+
+	documentFromPolicy := GetIamPolicyDocument(t, region, policyARN)
```
[the remark this answers is not in the document]
Comment (Briansbum) [reply modules/aws/iam.go]: 👍 will change

File: modules/aws/iam.go [hunk 3]
```diff
@@ -133,6 +138,97 @@ func EnableMfaDeviceE(t *testing.T, iamClient *iam.IAM, mfaDevice *iam.VirtualMF
...
+
+	assert.JSONEq(
+		t,
+		dst.String(),
+		documentFromPolicy,
+	)
```
[the remark this answers is not in the document]
Comment (Briansbum) [reply modules/aws/iam.go]: 👍 will change

File: modules/aws/iam.go [hunk 4]
```diff
@@ -133,6 +138,97 @@ func EnableMfaDeviceE(t *testing.T, iamClient *iam.IAM, mfaDevice *iam.VirtualMF
...
+	}
+
+	document, err := iamClient.GetPolicyVersion(&iam.GetPolicyVersionInput{
+		PolicyArn: aws.String(policyARN),
+		VersionId: aws.String(defaultVersion),
+	})
+
+	var unescapedDocument *string
+
+	unescapedDocument = document.PolicyVersion.Document
```
[the remark this answers is not in the document]
Comment (Briansbum) [reply modules/aws/iam.go]: I was having issues here with that way, when the `document.PolicyVersion.Document` came back as nil I'd get a nil pointer dereference. This is how I stopped that from happening.

File: modules/aws/iam.go [hunk 5]
```diff
@@ -133,6 +138,97 @@ func EnableMfaDeviceE(t *testing.T, iamClient *iam.IAM, mfaDevice *iam.VirtualMF
...
+
+	if unescapedDocument == nil {
+		return "", errors.New("there isn't a policy document in the default policy version")
```
[the remark this answers is not in the document]
Comment (Briansbum) [reply modules/aws/iam.go]: 👍 will change

File: modules/aws/iam.go [hunk 6]
```diff
@@ -133,6 +138,97 @@ func EnableMfaDeviceE(t *testing.T, iamClient *iam.IAM, mfaDevice *iam.VirtualMF
...
+// AssertIamPolicyExists checks if the given IAM policy exists in the given region and fail the test if it does not.
+func AssertIamPolicyExists(t *testing.T, region string, policyARN string) {
+	_, err := GetIamPolicyDocumentE(t, region, policyARN)
+	if err != nil {
+		t.Fatal(err)
+	}
+}
+
+// AssertIAMPolicyIsCorrect fetches the contents of the bucket policy for the given bucket and matches it against the provided policy document.
+func AssertIAMPolicyIsCorrect(t *testing.T, region string, policyARN string, policyDocument string) {
```
[the remark this answers is not in the document]
Comment (Briansbum) [reply modules/aws/iam.go]: Sure that'll work

File: modules/aws/iam_test.go [hunk 7]
```diff
@@ -19,3 +23,28 @@ func TestGetIamCurrentUserArn(t *testing.T) {
 	username := GetIamCurrentUserArn(t)
 	assert.Regexp(t, "^arn:aws:iam::[0-9]{12}:user/.+$", username)
 }
+
+func TestAssertIAMPolicyIsCorrect(t *testing.T) {
+	t.Parallel()
+
+	region := GetRandomRegion(t, nil, nil)
+
+	policyDocument := "{\n  \"Version\": \"2012-10-17\",\n  \"Statement\": [\n    {\n      \"Sid\": \"Stmt1530709892083\",\n      \"Action\": \"*\",\n      \"Effect\": \"Allow\",\n      \"Resource\": \"*\"\n    }\n  ]\n}"
```
[the remark this answers is not in the document]
Comment (Briansbum) [reply modules/aws/iam_test.go]: This format is how the API returns it but I would be surprised if `assert` can't handle comparing both formats so I'll give it a go.

File: modules/aws/s3.go [hunk 8]
```diff
@@ -161,6 +163,67 @@ func AssertS3BucketExistsE(t *testing.T, region string, name string) error {
...
+
+	policyFromBucket := GetS3BucketPolicyContents(t, region, bucket)
+
+	assert.JSONEq(
+		t,
+		dst.String(),
+		policyFromBucket,
+	)
+
+	return nil
```
[the remark this answers is not in the document]
Comment (Briansbum) [reply modules/aws/s3.go]: 👍 will change

File: modules/aws/s3_test.go [hunk 9]
```diff
@@ -50,3 +52,33 @@ func TestAssertS3BucketExistsNoFalsePositive(t *testing.T) {
 		t.Fatalf("Function claimed that S3 Bucket '%s' exists, but in fact it does not.", s3BucketName)
 	}
 }
+
+func TestAssertS3BucketPolicyIsCorrect(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (Briansbum) [reply modules/aws/s3_test.go]: 😆

File: modules/aws/s3_test.go [hunk 10]
```diff
@@ -50,3 +52,33 @@ func TestAssertS3BucketExistsNoFalsePositive(t *testing.T) {
...
+	t.Parallel()
+
+	region := GetRandomRegion(t, nil, nil)
+	s3BucketName := "gruntwork-terratest-" + strings.ToLower(random.UniqueId())
+	logger.Logf(t, "Random values selected. Region = %s, s3BucketName = %s\n", region, s3BucketName)
+
+	CreateS3Bucket(t, region, s3BucketName)
+	defer DeleteS3Bucket(t, region, s3BucketName)
```
[the remark this answers is not in the document]
Comment (Briansbum) [reply modules/aws/s3_test.go]: 👍 will change

File: modules/aws/iam.go [hunk 11]
```diff
@@ -135,26 +140,43 @@ func EnableMfaDeviceE(t *testing.T, iamClient *iam.IAM, mfaDevice *iam.VirtualMF
...
+	_, err := GetIamPolicyDocumentE(t, region, policyARN)
 	if err != nil {
 		t.Fatal(err)
 	}
 }
 
-//AssertIamPolicyExistsE checks if the given IAM policy exists in the given region and return an error if it does not.
-func AssertIamPolicyExistsE(t *testing.T, region string, policyARN string) error {
-	iamClient, err := NewIamClientE(t, region)
+// AssertIAMPolicyIsCorrect fetches the contents of the bucket policy for the given bucket and matches it against the provided policy document.
```
Comment (codevbus) [inline modules/aws/iam.go]: This seems a bit confusing. IAM policy != S3 Bucket policy.

File: modules/aws/iam.go [hunk 12]
```diff
@@ -135,26 +140,43 @@ func EnableMfaDeviceE(t *testing.T, iamClient *iam.IAM, mfaDevice *iam.VirtualMF
...
+func AssertIAMPolicyIsCorrect(t *testing.T, region string, policyARN string, policyDocument string) {
+	err := AssertIAMPolicyIsCorrectE(t, region, policyARN, policyDocument)
+	if err != nil {
+		t.Fatal(err)
+	}
+}
+
+// AssertIAMPolicyIsCorrectE fetches the contents of the bucket policy for the given bucket and matches it against the provided policy document returning an error if they don't match.
```
Comment (codevbus) [inline modules/aws/iam.go]: Same comment as: https://github.com/gruntwork-io/terratest/pull/130/commits/b958b3de6adc8c4884993a8f2739ef3254dad7ef#r201366000

File: modules/aws/iam.go (hunk 11 above)
Comment (Briansbum) [reply modules/aws/iam.go]: todays lesson, don't copy-paste so much

File: modules/aws/iam_test.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (Briansbum) [reply modules/aws/iam_test.go]: Assert didn't like it sadly, it's messy but I think this has to stay.

File: modules/aws/s3.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (Briansbum) [reply modules/aws/s3.go]: I wasn't sure where to put that function so it's inside of `iam.go` for the moment.

**Comment by Briansbum** [comment]
@brikis98 Sorry that this has taken so long, it's now rebased against master.

File: examples/terraform-aws-example/README.md [hunk 13]
```diff
@@ -25,7 +25,7 @@ it should be free, but you are completely responsible for all AWS charges.
 1. Sign up for [AWS](https://aws.amazon.com/).
 1. Configure your AWS credentials using one of the [supported methods for AWS CLI
    tools](https://docs.aws.amazon.com/cli/latest/userguide/cli-chap-getting-started.html), such as setting the
-   `AWS_ACCESS_KEY_ID` and `AWS_SECRET_ACCESS_KEY` environment variables. If you're using the `~/.aws/config` file for profiles then export `AWS_SDK_LOAD_CONFIG` as "True".
```
Comment (brikis98) [inline examples/terraform-aws-example/README.md]: Why remove this?

File: modules/aws/ami.go [hunk 14]
```diff
@@ -26,7 +26,7 @@ func DeleteAmiAndAllSnapshots(t *testing.T, region string, ami string) {
 	}
 }
 
-// DeleteAmiAndAllSnapshotsE will delete the given AMI along with all EBS snapshots that backed that AMI
```
Comment (brikis98) [inline modules/aws/ami.go]: This also seems strange... Did you rebase or merge incorrectly?

File: modules/aws/iam.go [hunk 15]
```diff
@@ -1,14 +1,60 @@
...
+	dst := new(bytes.Buffer)
+
+	src := []byte(correct)
+
+	err := json.Compact(dst, src)
+	if err != nil {
+		return err
+	}
+
+	equal := assert.JSONEq(
```
Comment (brikis98) [inline modules/aws/iam.go]: This will fail the test it they are not equal, which violates the contract of the methods that end in `E`. You'll need to do the JSON comparison using some other approach.

File: modules/aws/iam.go [hunk 16]
```diff
@@ -1,14 +1,60 @@
...
+		t,
+		dst.String(),
+		comparison,
+	)
+	if equal == false {
+		return errors.New("text")
```
Comment (brikis98) [inline modules/aws/iam.go]: Need a better error message than "text"!

File: modules/aws/iam.go [hunk 17]
```diff
@@ -133,6 +179,90 @@ func EnableMfaDeviceE(t *testing.T, iamClient *iam.IAM, mfaDevice *iam.VirtualMF
...
+}
+
+// AssertIAMPolicyIsEqualE fetches the contents of the IAM policy and matches it against the provided policy document returning an error if they don't match.
+func AssertIAMPolicyIsEqualE(t *testing.T, region string, policyARN string, policyDocument string) error {
+	documentFromPolicy, err := GetIamPolicyDocumentE(t, region, policyARN)
+	if err != nil {
+		return err
+	}
+
+	err = AssertJsonEqualE(t, policyDocument, documentFromPolicy)
```
Comment (brikis98) [inline modules/aws/iam.go]: You can just `return AssertJsonEqualE(...)`

File: modules/aws/iam.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (brikis98) [reply modules/aws/iam.go]: That wouldn't be due to a variable declaration... Perhaps you were derferencing it before? i.e., `unescapedDocument = *document.PolicyVersion.Document`

File: modules/aws/iam_test.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (brikis98) [reply modules/aws/iam_test.go]: Shouldn't `AssertIAMPolicyIsEqual` handle whitespace differences?

File: modules/aws/s3_test.go [hunk 18]
```diff
@@ -107,11 +137,10 @@ func TestEmptyS3BucketVersioned(t *testing.T) {
 	}
 
 	_, err = s3Client.PutBucketVersioning(versionInput)
+
 	if err != nil {
 		t.Fatal(err)
 	}
-
-	testEmptyBucket(t, s3Client, region, s3BucketName)
```
Comment (brikis98) [inline modules/aws/s3_test.go]: Why is this removed?

File: modules/aws/s3.go [hunk 19]
```diff
@@ -102,7 +102,7 @@ func CreateS3Bucket(t *testing.T, region string, name string) {
 
 // CreateS3BucketE creates an S3 bucket in the given region with the given name. Note that S3 bucket names must be globally unique.
 func CreateS3BucketE(t *testing.T, region string, name string) error {
-	logger.Logf(t, "Creating bucket %s in %s", name, region)
+	logger.Logf(t, "Creating bucket %s in %s", region, name)
```
Comment (liquid-sky) [inline modules/aws/s3.go]: Variable order needs to be swapped.

**Comment by codevbus** [comment]
@Briansbum - still working on this? I'd love to see the feature set of this tool expand, specifically with regards to IAM.

**Comment by barundel** [comment]
would love to see this :)

**Comment by devopsinfoltd** [comment]
@Briansbum Any update mate on this? Really looking something like this

**Comment by ellisonc** [comment]
@Briansbum, thanks for your contribution. Unfortunately we’ve fallen behind on PRs and are removing old PRs. If this change is still required, we encourage you to re-open.
